### PR TITLE
Implement onRefreshAttempt callback, setters for refresh/subscribe mechanism

### PR DESCRIPTION
**Pull request by ernsheong**
Closes https://github.com/centrifugal/centrifuge-js/issues/86

### CODE REVIEW

File: src/centrifuge.js [hunk 1]
```diff
@@ -105,6 +105,26 @@ export class Centrifuge extends EventEmitter {
...
+  }
+
+  setSubscribeHeaders(headers) {
+    this._config.subscribeHeaders = headers;
+  }
+  setSubscribeParams(params) {
+    this._config.subscribeParams = params;
+  }
+  setSubscribeData(data) {
+    this._config.subscribeData = data;
```
Comment (FZambia) [inline src/centrifuge.js]: There is no `subscribeData` field in config

File: src/centrifuge.js [hunk 2]
```diff
@@ -735,11 +755,17 @@ export class Centrifuge extends EventEmitter {
           if (resolveCtx.next) {
             resolveCtx.next();
           }
+          if (this._config.onRefreshAttempt) {
+            this._config.onRefreshAttempt(true, resolveCtx);
```
Comment (FZambia) [inline src/centrifuge.js]: we need to create custom new context here that will only include information we want to export - at moment `resolveCtx` can contain some internal fields.

File: src/centrifuge.js [hunk 3]
```diff
@@ -735,11 +755,17 @@ export class Centrifuge extends EventEmitter {
...
+          }
         }, rejectCtx => {
           this._refreshError(rejectCtx.error);
           if (rejectCtx.next) {
             rejectCtx.next();
           }
+          if (this._config.onRefreshAttempt) {
+            this._config.onRefreshAttempt(false, rejectCtx);
```
Comment (FZambia) [inline src/centrifuge.js]: the same note about context here - we need to create new object with required exported fields only.

**Comment by FZambia** [comment]
@ernsheong thanks! Added some review comments.

File: src/centrifuge.js (hunk 3 above)
Comment (ernsheong) [reply src/centrifuge.js]: Updated, see if this is what you meant

File: src/centrifuge.js (hunk 2 above)
Comment (ernsheong) [reply src/centrifuge.js]: Updated

**Comment by ernsheong** [comment]
@FZambia Is it better now? Feel free to change my branch

**Comment by FZambia** [comment]
@ernsheong yeah, this is better. One more thing is that new functionality should be added to [type definitions](https://github.com/centrifugal/centrifuge-js/blob/master/dist/centrifuge.d.ts). Also I don't remember at moment what `resolveCtx.result` contains - need to check its fields.

**Comment by ernsheong** [comment]
I just noticed the presence of `onRefresh` in the code, which is undocumented in README. Should we be using that instead? It's not clear to me what this is supposed to be for. 

CHANGELOG: https://github.com/centrifugal/centrifuge-js/blob/master/CHANGELOG.md#147

**Comment by FZambia** [comment]
Yeah, that callback allows to fully control refresh on application side. Using it you will have to construct and send Ajax request yourself. If this fits you we can consider to properly document it instead.

**Comment by ernsheong** [comment]
@FZambia `resolveCtx.result` is from below:
image

**Comment by FZambia** [comment]
@ernsheong I will look very soon - at moment a bit busy at work. So you decided that `onRefresh` does not fit your case, right?

**Comment by ernsheong** [comment]
`onRefresh` fits my case, just needs more documentation and also some convenience setters for headers and we will be good to go. I will close this PR for now in favor of another PR.

**Comment by FZambia** [comment]
@ernsheong just added a proper [documentation about onRefresh](https://github.com/centrifugal/centrifuge-js#refreshendpoint) callback

**Comment by ernsheong** [comment]
Looks good 👍

**Pull request closed without merging**